pxr/usd/usdSemantics/labelsQuery.cpp

```diff
@@ -0,0 +1,232 @@
+    if (!schema) {
+        return false;
+    }
+
+    auto doubleCheckedLockCondition = [&prim, this]() {
+        return _cachedLabels.count(prim.GetPath()) > 0;
+    };
+    {
+        std::shared_lock lock{_cachedLabelsMutex};
+        if (doubleCheckedLockCondition()) {
+            return true;
+        }
+    }
+
+    // Avoid locking while we compute
+    auto labels = std::visit([&schema](auto&& queryTime) -> _UnorderedTokenSet{
+        using TimeType = std::decay_t<decltype(queryTime)>;
+        const auto labelsAttr = schema.GetLabelsAttr();
+        if (!labelsAttr) {
+            TF_RUNTIME_ERROR("Labels attribute undefined at %s",
+                             UsdDescribe(schema.GetPrim()).c_str());
+            return {};
+        }
+        if constexpr (std::is_same_v<TimeType, UsdTimeCode>) {
+            VtTokenArray labelsAtTime;
+            if (!labelsAttr.Get(&labelsAtTime, queryTime)) {
+                TF_RUNTIME_ERROR(
+                    "Failed to read tokens from %s",
+                    UsdDescribe(labelsAttr).c_str());
+            }
+            return {std::cbegin(labelsAtTime), std::cend(labelsAtTime)};
+        }
+        if constexpr (std::is_same_v<TimeType, GfInterval>) {
+            TF_DEV_AXIOM(!queryTime.IsEmpty());
+            std::vector<double> times;
+            if (!labelsAttr.GetTimeSamplesInInterval(
+                queryTime, &times)) {
+                TF_RUNTIME_ERROR("Failed to retrieve time samples at %s",
+                                 UsdDescribe(labelsAttr).c_str());
+                return {};
+            }
+            if (!TF_VERIFY(std::all_of(std::cbegin(times), std::cend(times),
+                           [](const auto t) {
+                                return t >= UsdTimeCode::EarliestTime(); }))
+                           ) {
+                return {};
+            }
+            // Ensure that the queryTime.GetMin() is considered if finite.
+            // If it's not, add the earliest time sample.
+            const double earliest =
+                queryTime.IsMinFinite() ? queryTime.GetMin() :
+                UsdTimeCode::EarliestTime().GetValue();
+            if (times.empty() || (times.front() != earliest)) {
+                times.push_back(earliest);
+            }
+
+            // Times is guaranteed to have interval.GetMin() or
+            // UsdTimeCode::EarliestTime.  If no time samples are authored,
+            // querying at this time sample will return default or fallback
+            // values. No special handling required.
+            TF_DEV_AXIOM(!times.empty());
+
+            _UnorderedTokenSet uniqueLabels;
+            for (const auto timeInInterval : times) {
+                VtTokenArray labelsAtTime;
+                if (!labelsAttr.Get(&labelsAtTime, timeInInterval)) {
+                    TF_RUNTIME_ERROR("Failed to read value at %s",
+                                     UsdDescribe(labelsAttr).c_str());
```

Comment:
I'm curious when we would hit the two conditions here that cause runtime errors to be emitted. Are those really runtime conditions that make that happen?

(Sorry to be so focused on error policy here, rather than what I'm supposed to be weighing in on!!)

Reply:
Thanks for doing so, @nvmkuruc , as I was just about to.  RUNTIME_ERRORs should be reserved for system-level errors.  Here and elsewhere, if we really expected there should have been something authored but there wasn't, it would be more of a TF_WARN, I think?

Reply:
Right, so that's "bad scene description", similar to having something like
```
over "mySphere"
{
    string radius = "five"
}
```
which is not considered a runtime error.  In Presto (and future USD) it would be a validation error, but at scene-interrogation-time, I think we are silent about it, actually.